src/Features/CSharp/Portable/Completion/CompletionProviders/ExplicitInterfaceMemberCompletionProvider.cs

```diff
@@ -5,133 +5,204 @@
+
+        var semanticModel = await newDocument.GetSemanticModelAsync(cancellationToken)
+            .ConfigureAwait(false);
+        cancellationToken.ThrowIfCancellationRequested();
+        Debug.Assert(semanticModel != null);
+
+        var baseMemberInterfaceType = member.ContainingType;
+        Debug.Assert(baseMemberInterfaceType.TypeKind is TypeKind.Interface);
 
-            var syntaxTree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+        var interfaceNode = await GetInterfaceNodeInCompletionAsync(newDocument, completionItem, baseMemberInterfaceType, cancellationToken).ConfigureAwait(false);
+        cancellationToken.ThrowIfCancellationRequested();
+        Debug.Assert(interfaceNode != null);
 
-            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
-            var semanticFacts = document.GetRequiredLanguageService<ISemanticFactsService>();
+        var state = AbstractImplementInterfaceService.State.Generate(implementInterfaceService, newDocument, semanticModel, interfaceNode, cancellationToken);
+        cancellationToken.ThrowIfCancellationRequested();
+        Debug.Assert(state != null);
 
-            if (syntaxFacts.IsInNonUserCode(syntaxTree, position, cancellationToken) ||
-                syntaxFacts.IsPreProcessorDirectiveContext(syntaxTree, position, cancellationToken))
+        var options = await newDocument.GetImplementTypeOptionsAsync(cancellationToken).ConfigureAwait(false);
+        var implementedSymbol = await implementInterfaceService.ExplicitlyImplementSingleInterfaceMemberAsync(
+            newDocument, state, member, options, cancellationToken)
+                .ConfigureAwait(false);
+
+        return implementedSymbol;
+    }
+
+    private static async Task<SyntaxNode?> GetInterfaceNodeInCompletionAsync(Document document, CompletionItem item, INamedTypeSymbol baseMemberInterfaceType, CancellationToken cancellationToken)
+    {
+        var syntaxTree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+        cancellationToken.ThrowIfCancellationRequested();
+        var documentSemanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+        cancellationToken.ThrowIfCancellationRequested();
+        Debug.Assert(documentSemanticModel != null, "Expected a semantic model out of the document");
+
+        var compilation = documentSemanticModel.Compilation;
+        var node = syntaxTree.FindNode(item.Span, false, true, cancellationToken);
+        var primaryTypeDeclaration = node.GetAncestor<BaseTypeDeclarationSyntax>();
+        Debug.Assert(primaryTypeDeclaration != null, "Expected a BaseTypeDeclarationSyntax to contain the implemented interface member");
+        var interfaceNode = NodeInDeclaration(primaryTypeDeclaration);
+        if (interfaceNode != null)
+        {
+            return interfaceNode;
+        }
```

Comment:
prior to this it would be good to doc the general idea of what's going on.  what you're looking for and what you expect the shape of the tree to be.

Reply:
Wrote docs, please check